### DISCUSSION TITLE
Type-safe custom actions

### DISCUSSION THREAD

**Discussion by margalit** [wevm/wagmi#1538] [Q&A] [answered]
More of a TypeScript question than wagmi-specific but I'm wondering if anyone can help me create a custom version of an action that pre-fills some contract defaults but still has type-safety around functionName and args? e.g:

```typescript
function readWagmigotchi(config: <~ Help me type this badboy) {
  return readContract({
    address: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
    abi: wagmigotchiABI,
    ...config
  })
}
```
I've tried variations of `ReadContractConfig<typeof wagmigotchiABI, TFunctionName>` where `TFunctionName` is a generic captured by readWagmigotchi but no luck so far.

I'm sure I could extract the correct types by using `abitype` but wondering if there's a solution with the types wagmi + @wagmi/core already export.

**Comment by tmm** [comment, marked as answer]
Have an upcoming feature that will help out with this, but until then, here's an example ([TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nASighgEwGEIA7GKXAYxmJIDNgBzOAXzjqghDgHIABAO65GIYAHpK0FDwBQoSLAxwoeIqXJUYAGjgoolAEwAGAIIAhAJKt2nbvyEjxklbJl0AriWrBSy1QHVhUUYIGEoAC2AAHhk4OAAVEwAjYF0ADxgUEnwAZz8CUgAbRDhPAGsSCAESAG0AXTgAXgRkFAg6XX1jcwstWISAMU9vUgA5XBA0FAys3Lgc8mASZmaFqCXGPoA+AApJeiYALjgAeVEYKJwCWg1qWgZGKMSUnXihrxgfEnHJrZ0eXApHhbACUGH6Khg7igJHyajIFGoO3Q-TigOAxz0hlMlj6cTiADoifsHnjrLg8uVKtU4BTsKoboiaKQHk9ksBXu8Rt8JihQTIWDIZPsFn48s1cEJgPAVARAo4QmFIsj+gR8CocjljjwjGlAGQEPDJHg+Xx+KG1SVwhVwXhQJzohv6YjEcAAegB+VVQRhauA1HX6nh1MnOt2elggmQamShj1AA)):

```ts
import type { ReadContractConfig } from '@wagmi/core'
import { readContract, erc20ABI } from '@wagmi/core'

function readWagmigotchi<
  TAbi extends readonly unknown[] = typeof erc20ABI,
  TFunctionName extends string = string,
>(config: Omit<ReadContractConfig<TAbi, TFunctionName>, 'abi'>) {
  return readContract({
    abi: erc20ABI,
    ...config,
  } as unknown as ReadContractConfig<TAbi, TFunctionName>)
}

const res = await readWagmigotchi({
  address: '0x…',
  functionName: 'balanceOf',
  // ^? (property) functionName: "symbol" | "name" | "balanceOf" | "allowance" | "decimals" | "totalSupply"
  args: ['0x…'],
  // ^? (property) args: readonly [`0x${string}`]
})
res
// ^? const res: BigNumber
```

**Reply by margalit** [reply]
Fantastic, thanks @tmm!
